crates/telio-wg/src/no_link_detection.rs

```diff
@@ -0,0 +1,524 @@
+            (self.last_tx_bytes, self.last_tx_timestamp)
+        };
+
+        let link_state = if tx_ts <= rx_ts || rx_ts.elapsed() < WG_KEEPALIVE + rtt {
+            LinkState::Up
+        } else {
+            LinkState::Down
+        };
+
+        NodeInfo {
+            last_tx_timestamp: tx_ts,
+            last_rx_timestamp: rx_ts,
+            last_tx_bytes: tx_bytes,
+            last_rx_bytes: rx_bytes,
+            last_link_state: link_state,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::no_link_detection;
+
+    use super::*;
+    use tokio::time;
+
+    use telio_model::{api_config::FeatureNoLinkDetection, mesh::Node};
+
+    struct Env {
+        key: PublicKey,
+        nld: NoLinkDetection,
+    }
+
+    impl Env {
+        fn new() -> Self {
+            let features = FeatureNoLinkDetection { rtt_seconds: None };
+
+            Env {
+                key: PublicKey::default(),
+                nld: NoLinkDetection::new(Some(features)),
+            }
+        }
+    }
+
+    fn check_node_info(
+        info: &NodeInfo,
+        rx: u64,
+        tx: u64,
+        rx_ts: Instant,
+        tx_ts: Instant,
+        link_state: LinkState,
+    ) {
+        assert_eq!(info.last_rx_timestamp, rx_ts);
+        assert_eq!(info.last_tx_timestamp, tx_ts);
+
+        assert_eq!(info.last_rx_bytes, rx);
+        assert_eq!(info.last_tx_bytes, tx);
+
+        assert_eq!(info.last_link_state, link_state);
+    }
+
+    #[test]
+    fn test_config() {
+        // Disabled
+        let no_link_detection = NoLinkDetection::new(None);
+        assert_eq!(no_link_detection.is_disabled(), true);
+        assert_eq!(no_link_detection.rtt(), DEFAULT_CONFIGURABLE_RTT);
+
+        // Default RTT
+        let features = FeatureNoLinkDetection { rtt_seconds: None };
+        let no_link_detection = NoLinkDetection::new(Some(features));
+        assert_eq!(no_link_detection.is_disabled(), false);
+        assert_eq!(no_link_detection.rtt(), DEFAULT_CONFIGURABLE_RTT);
+
+        // Custom RTT
+        let features = FeatureNoLinkDetection {
+            rtt_seconds: Some(1),
+        };
+        let no_link_detection = NoLinkDetection::new(Some(features));
+        assert_eq!(no_link_detection.is_disabled(), false);
+        assert_eq!(no_link_detection.rtt(), Duration::from_secs(1));
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn test_insert() {
+        let Env { key, mut nld } = Env::new();
+        nld.insert(&key, None, None, None);
+        check_node_info(
+            nld.info.get(&key).unwrap(),
+            0,
+            0,
+            Instant::now(),
+            Instant::now(),
+            LinkState::Down,
+        );
+
+        time::advance(Duration::from_secs(1)).await;
+
+        nld.insert(&key, None, None, Some(LinkState::Up));
+        check_node_info(
+            nld.info.get(&key).unwrap(),
+            0,
+            0,
+            Instant::now(),
+            Instant::now(),
```

Comment:
What is the reason of advancing time in here? I assumed because we might call `update()` but we basically insert item into the table?